File: 0309-best-time-to-buy-and-sell-stock-with-cooldown/0309-best-time-to-buy-and-sell-stock-with-cooldown.cpp
class Solution {
public:
    int maxProfit(vector<int>& p) {
        int n = p.size(); //i
        if(n<=1) return 0;
        int selling = 0,  buying = -p[0], cooldown = 0;
        // no-stock - after sold or after cooldown state;
        for(int i=0;i<n;i++){
            int prv = selling ;
            selling = buying + p[i];
            buying = max( buying, cooldown - p[i]);
            cooldown = max(cooldown,prv);
        }
        return max(selling ,cooldown);
    }
};

File: 0309-best-time-to-buy-and-sell-stock-with-cooldown/0309-best-time-to-buy-and-sell-stock-with-cooldown.py
class Solution:
    def maxProfit(self, p: List[int]) -> int:
        n = len(p)
         
        dp = [[0]*2 for _ in range(n+2)]

        for i in range(n-1,-1,-1):                            
            dp[i][1] = max(-p[i] + dp[i+1][0],dp[i+1][1])                
            dp[i][0] = max(p[i] + dp[i+2][1], dp[i+1][0])
        
        return dp[0][1]